### PR TITLE
[bugfix](ngram bf index) process differently for normal bloom filter index and ngram bf index

### LINKED ISSUE #21234: [Bug] when using NGram index columns as query condition, be crash
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

doris-2.0.0-alpha1

### What's Wrong?

when using NGram index columns as query condition
1. if use `=` , be will  crash, eg:select * from t1 where col1='123' limit 10
2.  if use `like` , query ok  , eg:select * from t1 where col1 like '123' limit 10

be error log:
image

### What You Expected?

NGram index query ok 

### How to Reproduce?

_No response_

### Anything Else?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by xiaokang**
## Proposed changes

Issue Number: close #21234

Fix miss match of bloom filter index and query type. Use normal bloom filter index for eq query and ngram bloom filter index for like query.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xiaokang** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/column_reader.h [hunk 1]
```diff
@@ -137,7 +137,12 @@ class ColumnReader {
 
     bool has_zone_map() const { return _zone_map_index_meta != nullptr; }
     bool has_bitmap_index() const { return _bitmap_index_meta != nullptr; }
-    bool has_bloom_filter_index() const { return _bf_index_meta != nullptr; }
+    bool has_bloom_filter_index(bool ngram) const {
+        return _bf_index_meta != nullptr &&
+               (ngram ? (_bf_index_meta->algorithm() == BloomFilterAlgorithmPB::NGRAM_BLOOM_FILTER)
+                      : (_bf_index_meta->algorithm() !=
+                         BloomFilterAlgorithmPB::NGRAM_BLOOM_FILTER));
+    }
```
Comment (compasses) [inline be/src/olap/rowset/segment_v2/column_reader.h:145]: 这个判断有点复杂:)

**Comment by compasses** [comment]
LGTM

File: be/src/olap/null_predicate.h [hunk 2]
```diff
@@ -92,7 +94,7 @@ class NullPredicate : public ColumnPredicate {
         }
     }
 
-    bool can_do_bloom_filter() const override { return _is_null; }
+    bool can_do_bloom_filter(bool ngram) const override { return ngram ? false : _is_null; }
```
Comment (airborne12) [inline be/src/olap/null_predicate.h:97]: may be `bool can_do_bloom_filter(bool ngram) const override { return !ngram && _is_null; }` is better

File: be/src/olap/rowset/segment_v2/column_reader.h (hunk 1 above)
Comment (xiaokang) [reply be/src/olap/rowset/segment_v2/column_reader.h:148]: changed

**Comment by xiaokang** [comment]
run buildall

File: be/src/olap/in_list_predicate.h [hunk 3]
```diff
@@ -381,6 +381,8 @@ class InListPredicateBase : public ColumnPredicate {
 
     bool evaluate_and(const segment_v2::BloomFilter* bf) const override {
         if constexpr (PT == PredicateType::IN_LIST) {
+            // IN predicate can not use ngram bf, just return true to accept
+            if (bf->is_ngram_bf()) return true;
```
Comment (Gabriel39) [inline be/src/olap/in_list_predicate.h:385]: For delete condition, is this always true?

**Review by airborne12 — APPROVED** [review]
LGTM

**Comment by xiaokang** [comment]
run buildall

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 52.64 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17164968599 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230713100929_clickbench_pr_177611.html

**Pull request merged**